### ISSUE TITLE
Questions about noiseprint++

### ISSUE THREAD

**Issue by znb899** [grip-unina/TruFor#4]
Hello,

First, congratulations for your publication. Your work is very interesting and I'm particularly interesed in Noiseprint++
I tried to reproduce its output for one of the images in the paper but I don't get the same result. I'm guessing it may be a problem with normalizing the output to the range [0,1] that's not the same but I've tried other ways and it doesn't yield the same results.

In my case, although the jpeg grid artifacts can be clearly seen (which is impressive), the "semantics" of the image are still very visible.

```{python}
from models.dncnn import make_net
import torch
from collections import OrderedDict

state_dict_path = '../../weights/trufor.pth.tar'
state_dict = torch.load(state_dict_path)['state_dict']

new_state_dict = OrderedDict()
for u, v in state_dict.items():
    if u[:6] == 'dncnn.':
        new_state_dict[u[6:]] = v

num_levels = 17
out_channel = 1

model = make_net(3, kernels=[3, ] * num_levels,
                 features=[64, ] * (num_levels - 1) + [out_channel],
                 bns=[False, ] + [True, ] * (num_levels - 2) + [False, ],
                 acts=['relu', ] * (num_levels - 1) + ['linear', ],
                 dilats=[1, ] * num_levels,
                 bn_momentum=0.1, padding=1).cuda()

model.load_state_dict(new_state_dict)

import torch
import numpy as np
import matplotlib.pyplot as plt

import torchvision.transforms.functional as F
from torchvision.io import read_image, ImageReadMode

plt.rcParams["savefig.bbox"] = 'tight'


def show(imgs):
    if not isinstance(imgs, list):
        imgs = [imgs]
    fig, axs = plt.subplots(ncols=len(imgs), squeeze=False)
    for i, img in enumerate(imgs):
        img = img.detach()
        img = F.to_pil_image(img)
        axs[0, i].imshow(np.asarray(img))
        axs[0, i].set(xticklabels=[], yticklabels=[], xticks=[], yticks=[])
    plt.show()

def preprc_imagenet_torch(x):
    mean = torch.Tensor([0.485, 0.456, 0.406]).to(x.device)
    std  = torch.Tensor([0.229, 0.224, 0.225]).to(x.device)
    x = (x-mean[None, :, None, None]) / std[None, :, None, None]
    return x

image = '../../data/tampered1.png'
image = (read_image(image, mode=ImageReadMode.RGB)[:,:700,:700].float()/255.).cuda()

pred = model(preprc_imagenet_torch(image))
pred_processed = pred[0].repeat(3, 1, 1)
pred_processed = pred_processed + pred_processed.min()
pred_processed = pred_processed / pred_processed.max()

show([image, pred_processed])
```
This is what I get.
![image](https://github.com/grip-unina/TruFor/assets/83007857/8acb8340-a13e-44af-a05c-031fd9f7a491)

Could you please point out the errors in my code?

Also, when do you think you will be releasing the code for training Noiseprint++? I can't figure out how you do the dataloading part.

I want to train in on another type of aquisition device (scanners) but documents usually have these huge areas that are completely uniform in color. No model can be expected to identify the source, let alone the position, of such a patch. Do you have in suggestions in how to overcome this obstacle?

Thanks for your help

**Comment by fabrizioguillaro** [comment]
We don't have a release date for the training code yet.

There are some mistakes in your code:

- in the first line of your min-max normalization should be pred-pred.min(), and not pred+pred.min()
- in our code, the pre-processing "preprc_imagenet_torch" is only applied to the input of the RGB branch of the CMX architecture. In the file builder_np_conf.py you can see that this pre-processing is not applied on the input of dncnn:
`
        if 'NP++' in self.mods:
            modal_x = self.dncnn(rgb)
            modal_x = torch.tile(modal_x, (3, 1, 1))
        else:
            modal_x = None

        if self.prepro is not None:
            rgb = self.prepro(rgb)`

- when we normalize the image between 0 and 1, we do /256.0 instead of /255.0
- when you do F.to_pil_image(img), I don't know if Pillow does something weird to the noiseprint

I advice to visualize the noiseprint like we do in the visualize.py script, which removes some of the border and samples the noiseprint++. Changing the step can change the visualization.
I will add some of these information in the readme.

**Comment by cxs-ux** [comment]
I can't download the weights now.Could you hlep me?

**Comment by fabrizioguillaro** [comment]
Our website is currently down because of maintenance at our university. You should try again in the next days
